x/claim/keeper/claim.go

```diff
@@ -126,81 +142,87 @@ func (k Keeper) GetClaimableAmountForAction(ctx sdk.Context, addr string, action
-		claimableCoins = claimableCoins.Add(sdk.NewCoin(coin.Denom, claimableAmount))
-	}
+	claimableAmount := initalClaimableAmount.Amount.Mul(claimablePercent.Mul(sdk.NewDec(10000)).RoundInt()).QuoRaw(10000)
+	claimableCoin := sdk.NewCoin(initalClaimableAmount.Denom, claimableAmount)
 
-	return claimableCoins, nil
+	return claimableCoin, nil
 }
 
-// // ClaimCoins remove claimable amount entry and transfer it to user's account
-// func (k Keeper) ClaimCoinsForAction(ctx sdk.Context, addr sdk.AccAddress, action types.Action) (sdk.Coins, error) {
-// 	claimableAmount, err := k.GetClaimableAmountForAction(ctx, addr, action)
-// 	if err != nil {
-// 		return claimableAmount, err
-// 	}
-
-// 	if claimableAmount.Empty() {
-// 		return claimableAmount, nil
-// 	}
-
-// 	claimRecord, err := k.GetClaimRecord(ctx, addr)
-// 	if err != nil {
-// 		return nil, err
-// 	}
-
-// 	err = k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, addr, claimableAmount)
-// 	if err != nil {
-// 		return nil, err
-// 	}
-
-// 	claimRecord.ActionCompleted[action] = true
-
-// 	err = k.SetClaimRecord(ctx, claimRecord)
-// 	if err != nil {
-// 		return claimableAmount, err
-// 	}
-
-// 	ctx.EventManager().EmitEvents(sdk.Events{
-// 		sdk.NewEvent(
-// 			types.EventTypeClaim,
-// 			sdk.NewAttribute(sdk.AttributeKeySender, addr.String()),
-// 			sdk.NewAttribute(sdk.AttributeKeyAmount, claimableAmount.String()),
-// 		),
-// 	})
-
-// 	return claimableAmount, nil
-// }
+// GetModuleAccountBalance gets the airdrop coin balance of module account
+func (k Keeper) GetModuleAccountAddress(ctx sdk.Context) sdk.AccAddress {
+	return k.accountKeeper.GetModuleAddress(types.ModuleName)
+}
+
+// GetModuleAccountBalance gets the airdrop coin balance of module account
+func (k Keeper) GetModuleAccountBalance(ctx sdk.Context) sdk.Coin {
+	moduleAccAddr := k.GetModuleAccountAddress(ctx)
+	params := k.GetParams(ctx)
+	return k.bankKeeper.GetBalance(ctx, moduleAccAddr, params.ClaimDenom)
+}
+
+// ClaimCoins remove claimable amount entry and transfer it to user's account
+func (k Keeper) ClaimCoinsForAction(ctx sdk.Context, addr string, action types.Action) (sdk.Coin, error) {
+	claimableAmount, err := k.GetClaimableAmountForAction(ctx, addr, action, types.ARKEO)
+	if err != nil {
+		return claimableAmount, err
+	}
+
+	if claimableAmount.IsNil() {
+		return claimableAmount, nil
+	}
+
+	claimRecord, err := k.GetClaimRecord(ctx, addr, types.ARKEO)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+	accountAddress, err := sdk.AccAddressFromBech32(addr)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+	err = k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, accountAddress, sdk.NewCoins(claimableAmount))
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
+	claimRecord.ActionCompleted[action] = true
+
+	err = k.SetClaimRecord(ctx, claimRecord)
+	if err != nil {
+		return claimableAmount, err
```

Comment:
if it failed to save claimrecord , I think we should `return sdk.Coin{},err` , otherwise people can claim it again , right?

Reply:
returning the claimableAmount here doesn't enable them to claim again as the err isn't nil, but yea, I see no reason to return the claimableAmount in that state.